docs(api-cli): restricting access to a function MTA-1488 (#121)
scaleway/docs-content: compute/functions/api-cli/restricting-access-to-a-function.mdx

---
title: Restricting access to a function
description: This page explains how to restrict access to a function
excerpt: |
  This page explains how to restrict access to a function
keywords: cloud computing, scaleway, serverless, functions, namespace, manage
dates:
  validation: 2021-05-26
  posted: 2021-05-26
---

> This page shows you how to restrict access to your function by making it private, thus preventing unauthorized launching of it. 

Reviewer: ```suggestion
This page shows you how to restrict access to your function by making it private. This prevents unauthorized launching of the function.
```

<Message type="requirement">

  - You have an account and are logged into the [Scaleway Console](https://console.scaleway.com)
  - You have [created a function](/compute/functions/how-to/create-a-function/)
  - You have generated a Scaleway [API key](https://www.scaleway.com/en/docs/console/my-project/how-to/generate-api-key/)
  - You have either `curl`, [Postman](https://www.postman.com/), [Talend API tester](https://chrome.google.com/webstore/detail/talend-api-tester-free-ed/aejoelaoggembcahagimdiliamlcdmfm) or any tool to send HTTP requests

</Message>

## Generating a JWT token

> Privacy works with JWT tokens. A JWT Token can be retrieved from the API-endpoint `GET /issue-jwt`. Depending on the parameters, a JWT token can be valid for either a function, a container, or a namespace:

Reviewer: ```suggestion
Privacy can be ensured via **J**SON **W**eb **T**okens. A JWT Token can be retrieved from the API-endpoint `GET /issue-jwt`. Depending on the parameters, a JWT token can be valid for either a function, a container, or a namespace:
```

* `/issue-jwt?namespace_id=1`: issues JWT valid for all functions inside namespace with ID 1.
* `/issue-jwt?function_id=1`: issues JWT valid only for function with id 1.

> <Message type="note">

Reviewer: ```suggestion
<Message type="tip">
```

> You may (optional) provide an expiration date (formatted `yyyy-mm-ddT00:00:00Z`) for the token. 

Reviewer: ```suggestion
You may optionally provide an expiration date (formatted `yyyy-mm-ddT00:00:00Z`) for the token. 
```

For example: `/issue-jwt?expires_at=2022-01-02T00:00:00Z&namespace_id=1` will generate a token, valid for all functions and containers inside Namespace with id 1, and this token will be valid until January 2nd 2022.

</Message>

The token will have the following claims:

```yaml
{
 "application_claim": [
   {
     "namespace_id": "string",
     "function_id": "string" // optional: id of function
   }
 ]
}
```

<Message type="important">

* Tokens are not stored by Scaleway and can not be retrieved if lost (but new tokens can be generated).

> * Token revocation is not yet supported, the best way to reset the tokens is to destroy and recreate the namespaces and all of its functions.

Reviewer: ```suggestion
* Token revocation is not yet supported. The best way to reset tokens is to destroy and recreate the namespaces and all of their functions.
```

</Message>

### Using curl

> Use the following commnand to issue a JWT token for your Namespace from the command line using cURL: 

Reviewer: ```suggestion
Use the following command to issue a JWT token for your Namespace from the command line using cURL: 
```

```sh
curl --location --request GET 'https://api.scaleway.com/functions/v1beta1/regions/fr-par/issue-jwt?namespace_id={your namespace id}&expires_at=2022-06-22T00%3A00%3A00Z' --header 'X-Auth-Token: {your scaleway API key'}
```

### Using an API Tester (Talend API Tester)

The following steps are designed for [Talend API tester](https://chrome.google.com/webstore/detail/talend-api-tester-free-ed/aejoelaoggembcahagimdiliamlcdmfm). 

1. Specify the `GET` method.
2. Enter the API URL. 
    * For Namespaces:
      ```
      https://api.scaleway.com/functions/v1beta1/regions/fr-par/issue-jwt?namespace_id={your namespace id}&expires_at={expiration date}
      ```
    * For Functions:
      ```
      https://api.scaleway.com/functions/v1beta1/regions/fr-par/issue-jwt?function_id={your function id}&expires_at={expiration date}
      ```
3. Add a `X-Auth-Token` header with your Scaleway API key as value.

> 4. Send the request

Author: ```suggestion
4. Send the request.
```

> 5. Copy the request

Reviewer: Is this right, or should it be "copy the response"?

Author: ```suggestion
5. Copy the response.
```

<Lightbox src="scaleway-talend-function.png" alt="" />

## Using your token

A **private** function observes this behaviour:

> * If a call is done without `SCW_FUNCTIONS_TOKEN` header, the call is rejected (Status Code 404)

Reviewer: ```suggestion
* If a call is made without the `SCW_FUNCTIONS_TOKEN` header, the call is rejected (Status Code 404)
```

> * If `SCW_FUNCTIONS_TOKEN` header is provided, the token is validated using a public key attached to the namespace.

Reviewer: ```suggestion
* If the `SCW_FUNCTIONS_TOKEN` header is provided, the token is validated using a public key attached to the namespace.
```
